modules/images/webp-uploads/load.php

```diff
@@ -704,3 +704,159 @@ function webp_uploads_update_rest_attachment( WP_REST_Response $response, WP_Pos
 	return rest_ensure_response( $data );
 }
 add_filter( 'rest_prepare_attachment', 'webp_uploads_update_rest_attachment', 10, 3 );
+
+/**
+ * Inspect if the current call to `wp_update_attachment_metadata()` was done from within the context
+ * of an edit to an attachment either restore or other type of edit, in that case we perform operations
+ * to save the sources properties, specifically for the `full` size image due this is a virtual image size.
+ *
+ * @since n.e.x.t
+ *
+ * @see wp_update_attachment_metadata()
+ *
+ * @param array $data The current metadata of the attachment.
+ * @param int   $attachment_id The ID of the current attachment.
+ * @return array The updated metadata for the attachment to be stored in the meta table.
+ */
+function webp_wp_update_attachment_metadata( $data, $attachment_id ) {
+
+	$trace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS, 10 );
+
+	foreach ( $trace as $element ) {
+		if ( ! isset( $element['function'] ) ) {
+			continue;
+		}
+
+		switch ( $element['function'] ) {
+			case 'wp_save_image':
+				// Right after an image has been edited.
+				return webp_uploads_backup_sources( $attachment_id, $data );
+			case 'wp_restore_image':
+				// When an image has been restored.
+				return webp_uploads_restore_image( $attachment_id, $data );
+		}
+	}
+
+	return $data;
+}
+
+add_filter( 'wp_update_attachment_metadata', 'webp_wp_update_attachment_metadata', 10, 2 );
+
+/**
+ * Before saving the metadata of the image store a backup values for the sources and file property
+ * those files would be used and deleted by the backup mechanism, right after the metadata has
+ * been updated. It removes the current sources property due once this function is executed
+ * right after an edit has taken place and the current sources are no longer accurate.
+ *
+ * @since n.e.x.t
+ *
+ * @param int   $attachment_id The ID representing the attachment.
+ * @param array $data The current metadata of the attachment.
+ * @return array The updated metadata for the attachment.
+ */
+function webp_uploads_backup_sources( $attachment_id, $data ) {
+	if ( ! isset( $data['sources'] ) ) {
+		return $data;
+	}
+
+	$target = isset( $_REQUEST['target'] ) ? sanitize_text_field( $_REQUEST['target'] ) : 'all';
+
+	// When an edit to an image is only applied to a thumbnail there's nothing we need to backu up.
+	if ( 'thumbnail' === $target ) {
+		return $data;
+	}
+
+	// Store the current sources before the current metadata is overwritten.
+	$sources             = get_post_meta( $attachment_id, '_wp_attachment_backup_sources', true );
+	$sources             = is_array( $sources ) ? $sources : array();
+	$sources['_sources'] = $data['sources'];
```

Comment:
> Again, I'm not sure I follow how this supports history. By storing the sources unconditionally in a hard-coded _sources, there will always ever be one _sources stored.

Basically, the way this works in WP is it updates the metadata of the image first and last it would update the backup metadata.

https://github.com/WordPress/wordpress-develop/blob/2648a5f984b8abf06872151898e3a61d3458a628/src/wp-admin/includes/image-edit.php#L981-L982

So at this point, we don't know if we have any backup already, so first we store the current sources for the full image before is updated then after the backup sizes is populated we retrieve those values and store it into the right meta.

The reason for this is that at this point (before saving the backup sizes) we don't know if we already have a `full-orig` or a `full-hash` value and we can't do that after either because at that point the metadata was already overwritten so instead we first store the values we need to preserve. 

Then after the values are stored in the backup we move it to the correct location here:

- https://github.com/WordPress/performance/blob/ba0cbdf45808d8844fbd371da66e70b57ff5cdd3/modules/images/webp-uploads/load.php#L854

> Also, why are we looking at $data here?

Great point here we should back up the current metadata as you described at this point the `$data` would be the new values. So let me update that one.